### PR TITLE
[Feature] Support LFB

### PR DESCRIPTION
**Pull request by congee524**
Support Long-Term Feature Bank.

### CODE REVIEW

File: configs/detection/lfb/i3d_for_lfb.py [hunk 1]
```diff
@@ -0,0 +1,142 @@
+# model settings
```
Comment (kennymckormick) [inline configs/detection/lfb/i3d_for_lfb.py:1]: What is this config for? If it is used as the pre-train weight for lfb detector, you need to tell the performance difference on Kinetics-400 and AVA between (1, 2, 2, 2) i3d and (1, 2, 2, 1) i3d.

File: configs/detection/lfb/i3d_for_lfb_affine_layer.py [hunk 2]
```diff
@@ -0,0 +1,140 @@
+# model settings
```
Comment (kennymckormick) [inline configs/detection/lfb/i3d_for_lfb_affine_layer.py:1]: Also this

File: configs/detection/lfb/lfb_ava.py [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+            lfb_prefix_path=lfb_prefix_path,
+            fbo_type='non_local',
+            fbo_cfg=dict(
+                num_non_local_layers=2,
+                num_st_feat_channels=2048,
+                num_lt_feat_channels=2048,
+                num_latent_channels=512)),
+        bbox_head=dict(
+            type='BBoxHeadAVA',
+            in_channels=2569,
```
Comment (kennymckormick) [inline configs/detection/lfb/lfb_ava.py:45]: Why 2569?

File: mmaction/models/common/affine_layer.py [hunk 4]
```diff
@@ -0,0 +1,25 @@
+import torch
+import torch.nn as nn
+from mmcv.cnn import NORM_LAYERS
+from torch.nn.parameter import Parameter
+
+
+@NORM_LAYERS.register_module()
+class AffineLayer(nn.Module):
```
Comment (kennymckormick) [inline mmaction/models/common/affine_layer.py:8]: By the way, I don't know why we need an AffineLayer. I think merging BN into Convs is OK for finetuning on AVA. Maybe we can study this problem after you got some performance numbers on AVA.

File: mmaction/models/heads/fbo_head.py [hunk 5]
```diff
@@ -0,0 +1,323 @@
...
+import mmcv
+import numpy as np
+import torch
+import torch.nn as nn
+from mmcv.cnn import ConvModule, constant_init, kaiming_init
+from mmcv.runner import auto_fp16, load_checkpoint
+
+try:
+    from mmdet.models.builder import SHARED_HEADS
+    from mmdet.utils import get_root_logger
```
Comment (kennymckormick) [inline mmaction/models/heads/fbo_head.py:15]: You should not import get_root_logger from mmdet.

File: mmaction/models/heads/fbo_head.py [hunk 6]
```diff
@@ -0,0 +1,323 @@
...
+except (ImportError, ModuleNotFoundError):
+    warnings.warn('Please install mmdet to use bbox2roi, SHARED_HEADS '
```
Comment (kennymckormick) [inline mmaction/models/heads/fbo_head.py:17]: Can update the log message.

File: mmaction/models/heads/lfb_head.py [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+try:
+    import mmdet  # noqa
+    from mmdet.core.bbox import bbox2roi
+    from mmdet.models import HEADS as MMDET_HEADS
+    from mmdet.models.roi_heads import StandardRoIHead
+except (ImportError, ModuleNotFoundError):
+    warnings.warn('Please install mmdet to use bbox2roi, MMDET_HEADS '
+                  'and StandardRoIHead')
+
+if 'mmdet' in dir():
```
Comment (kennymckormick) [inline mmaction/models/heads/lfb_head.py:16]: Better use the new checking style (after successfully importing all modules from mmdet, assign mmdet_imported as True). Cuz sometimes, you may successfully import mmdet, but not the submodules.

File: mmaction/models/heads/lfb_head.py [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+
+    @MMDET_HEADS.register_module()
+    class LFBRoIHead(StandardRoIHead):
+
+        def forward_dummy(self, x, proposals, img_metas):
```
Comment (kennymckormick) [inline mmaction/models/heads/lfb_head.py:21]: Where is `forward_dummy` used?

File: mmaction/models/heads/lfb_head.py [hunk 9]
```diff
@@ -0,0 +1,122 @@
+import warnings
```
Comment (kennymckormick) [inline mmaction/models/heads/lfb_head.py:1]: Maybe you can modify and directly use the AVARoIHead? Seems the only difference is that you also pass kwargs to the shared head. Maybe no need to add a `lfb_head`

Comment (congee524) [reply mmaction/models/heads/lfb_head.py:1]: yeah，then I don't need to import mmdet

File: configs/detection/lfb/lfb_slowonly_nl.py [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+                       'recall_93.9.pkl')
+proposal_file_val = f'{anno_root}/ava_dense_proposals_val.FAIR.recall_93.9.pkl'
+
+img_norm_cfg = dict(
+    mean=[123.675, 116.28, 103.53], std=[58.395, 57.12, 57.375], to_bgr=False)
+
+mc_cfg = dict(
+    server_list_cfg='/mnt/lustre/share/memcached_client/server_list.conf',
+    client_cfg='/mnt/lustre/share/memcached_client/client.conf',
+    sys_path='/mnt/lustre/share/pymc/py3')
```
Comment (dreamerlin) [inline configs/detection/lfb/lfb_slowonly_nl.py:56]: : P

File: mmaction/models/common/lfb.py [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+            1705: tensor([[-1.0169, -1.1293,  ...,  0.6793, -2.0540],
+                [ 1.2436, -0.4555,  ...,  0.2281, -0.8219],
+                [ 0.2815, -0.0547,  ..., -0.4199,  0.5157]]),
+            ...
+        },
+        'xmqSaQPzL1E': {
+            ...
+        },
+        ...
+    }
```
Comment (dreamerlin) [inline mmaction/models/common/lfb.py:36]: This should be in rst format, see RawFrameDataset or other places for reference

File: mmaction/models/common/lfb.py [hunk 12]
```diff
@@ -0,0 +1,130 @@
...
+            (including training and validation) occupies about 2GB GPU memory.
+            Default: 'gpu'.
+    """
+
+    def __init__(self,
+                 lfb_prefix_path,
+                 max_num_sampled_feat=5,
+                 window_size=60,
+                 lfb_channels=2048,
+                 dataset_modes=['train', 'val'],
```
Comment (dreamerlin) [inline mmaction/models/common/lfb.py:62]: Use tuple as default value. Remerber to change the docstring

File: configs/detection/lfb/lfb_slowonly_nl.py (hunk 10 above)
Comment (congee524) [reply configs/detection/lfb/lfb_slowonly_nl.py:56]: :p, could I keep it until the task finished? Otherwise, it's more troublesome to change every time

**Comment by congee524** [comment]
This PR has been almost ready. I will add unit tests next week. Maybe you could review it when you are free. :p

File: configs/detection/lfb/lfb_slowonly_infer.py [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+proposal_file_infer = (
+    f'{anno_root}/ava_dense_proposals_{dataset_mode}.FAIR.recall_93.9.pkl')
+
+img_norm_cfg = dict(
+    mean=[123.675, 116.28, 103.53], std=[58.395, 57.12, 57.375], to_bgr=False)
+
+mc_cfg = dict(
+    server_list_cfg='/mnt/lustre/share/memcached_client/server_list.conf',
+    client_cfg='/mnt/lustre/share/memcached_client/client.conf',
+    sys_path='/mnt/lustre/share/pymc/py3')
```
Comment (dreamerlin) [inline configs/detection/lfb/lfb_slowonly_infer.py:37]: here

File: configs/detection/lfb/lfb_max_kinetics_pretrained_slowonly_r50_4x16x1_20e_ava_rgb.py [hunk 14]
```diff
@@ -0,0 +1,141 @@
...
+label_file = f'{anno_root}/ava_action_list_v2.1_for_activitynet_2018.pbtxt'
+
+proposal_file_train = (f'{anno_root}/ava_dense_proposals_train.FAIR.'
+                       'recall_93.9.pkl')
+proposal_file_val = f'{anno_root}/ava_dense_proposals_val.FAIR.recall_93.9.pkl'
+
+img_norm_cfg = dict(
+    mean=[123.675, 116.28, 103.53], std=[58.395, 57.12, 57.375], to_bgr=False)
+
+mc_cfg = dict(
```
Comment (innerlee) [inline configs/detection/lfb/lfb_max_kinetics_pretrained_slowonly_r50_4x16x1_20e_ava_rgb.py:43]: here

File: configs/detection/lfb/lfb_max_kinetics_pretrained_slowonly_r50_4x16x1_20e_ava_rgb.py [hunk 15]
```diff
@@ -0,0 +1,141 @@
...
+    server_list_cfg='/mnt/lustre/share/memcached_client/server_list.conf',
+    client_cfg='/mnt/lustre/share/memcached_client/client.conf',
+    sys_path='/mnt/lustre/share/pymc/py3')
+
+train_pipeline = [
+    dict(type='SampleAVAFrames', clip_len=4, frame_interval=16),
+    dict(type='RawFrameDecode', io_backend='memcached', **mc_cfg),
```
Comment (innerlee) [inline configs/detection/lfb/lfb_max_kinetics_pretrained_slowonly_r50_4x16x1_20e_ava_rgb.py:50]: Use disk here

File: mmaction/models/common/lfb.py [hunk 16]
```diff
@@ -0,0 +1,187 @@
...
+        rank, world_size = get_dist_info()
+
+        # Loading LFB
+        if self.device == 'gpu':
+            self.load_lfb(f'cuda:{rank}')
+        elif self.device == 'cpu':
+            if world_size > 1:
+                warnings.warn(
+                    'If distributed training is used with multi-GPUs, lfb '
+                    "will be loaded repeatly on RAM. In this case, 'lmdb' is "
```
Comment (innerlee) [inline mmaction/models/common/lfb.py:101]: repeatly -> multiple times

File: mmaction/models/common/lfb.py [hunk 17]
```diff
@@ -0,0 +1,187 @@
...
+                    'more recomended.', UserWarning)
```
Comment (innerlee) [inline mmaction/models/common/lfb.py:102]: ~~more~~

**Comment by innerlee** [comment]
The code looks good

**Comment by congee524** [comment]
We don't have a unit test for detectors like `test_recognizers` and `test_localizers` right now, so maybe we need a PR to fix it.

**Comment by congee524** [comment]
This PR is ready.

**Review by innerlee — APPROVED** [review]

**Pull request merged**